### PR TITLE
Make `copymutable` public and add clarification to some docstrings

### LINKED ISSUE #51932: docs for `sort` refer to `Base.copymutable`, but the latter is not in the manual
> Uses `Base.copymutable` to support immutable collections and iterables.

If `Base.copymutable` is an internal function, why does the `sort` doc refer to it? Should `Base.copymutable` be made public?

### PR DESCRIPTION
**Pull request by LilithHafner**
Fixes #51932

- update sort docstring
- make copymutable public
- clarify that users should not typically need to override copymutable

### CODE REVIEW

File: base/sort.jl [hunk 1]
```diff
@@ -1474,7 +1474,9 @@ end
 
 Variant of [`sort!`](@ref) that returns a sorted copy of `v` leaving `v` itself unmodified.
 
-Uses `Base.copymutable` to support immutable collections and iterables.
+By default, `sort` copies its input with [`Base.copymutable`](@ref) before sorting, but some types have
+specialized versions of `sort` such as `NTuples` (`sort(::NTuple)` returns an `NTuple`) and
```
Comment (nsajko) [inline base/sort.jl:1478]: A comma is missing before "such as", and a period at the end of the sentence.

File: base/abstractarray.jl [hunk 2]
```diff
@@ -1201,6 +1201,9 @@ this is equivalent to `copy(a)`, but for other array types it may
 differ depending on the type of `similar(a)`.  For generic iterables
 this is equivalent to `collect(a)`.
 
+To change the behavior of `copymutable` on a custom collection is typically
+sufficient to define `similar` for that collection.
```
Comment (nsajko) [inline base/abstractarray.jl:1205]: xref `similar`?

**Comment by Tokazama** [comment]
Would it make sense to also include `emptymutable` in this?

**Comment by PallHaraldsson** [comment]
>make copymutable public

You're exporting, implicitly making public. It's probably ok, but should the original text say "export copymutable", be clear, or would it actually be made public, and NOT exported? I'm not worried in this case (about conflict), but I am in others, about exporting from Base.

**Comment by LilithHafner** [comment]
> Would it make sense to also include `emptymutable` in this?

It would make sense to use mark `emptymutable` as public if there are compelling reasons to include it in the public API. Let's leave that to another PR, though, as it is acceptable to have `copymutable` public but not `emptymutable`.

> You're exporting, implicitly making public.

No. I'm explicitly marking it as public. Please check the diff.

**Comment by PallHaraldsson** [comment]
>No. I'm explicitly marking it as public.

Sorry for the noise, yes PR and description matches (my intention was to alert to syncing, and it's good to know you're NOT exporting, and things match).

I saw you editing "base/exports.jl" and missed seeing "public", so I assumed. It would be helpful to have separate file for that base/public.jl, which could import base/exports.jl.

[You were editing line 1117 and public not visible until I expand/scroll up to line 1075, and it's not marked red as other keywords...]

File: base/abstractarray.jl (hunk 2 above)
Comment (LilithHafner) [reply base/abstractarray.jl:1205]:
```suggestion
sufficient to define [`similar`](@ref) for that collection.
```

**Comment by LilithHafner** [comment]
@nsajko is it more clear now?

**Comment by jakobnissen** [comment]
Just some minor comments:
* The docstring mentions that it's like `collect`, except for some other array types. I think that's not right - for non-array types it's certainly not equivalent to `collect`:
```julia
julia> s = Set([1,2,3]);

julia> s2 = Base.copymutable(s);

julia> typeof(s2)
Set{Int64}

julia> s === s2
false
```
* Maybe the example could use a Base type, and not require loading an external package?

File: base/sort.jl [hunk 3]
```diff
@@ -1474,7 +1474,9 @@ end
 
 Variant of [`sort!`](@ref) that returns a sorted copy of `v` leaving `v` itself unmodified.
 
-Uses `Base.copymutable` to support immutable collections and iterables.
+By default, `sort` copies its input with [`Base.copymutable`](@ref) before sorting, but some types have
+specialized versions of `sort`, such as `NTuples` (`sort(::NTuple)` returns an `NTuple`) and
+`Dict` (`sort(::Dict)` produces a new sorted dictionary data structure).
```
[the remark this answers is not in the document]
Comment (rfourquet) [reply base/sort.jl:1479]: In other words, while `copymutable` seems fine for arrays, I wonder whether `collect` wouldn't be better than `copymutable` for other iterables. As `copymutable` for arrays is just a thin wrapper atop `similar`, mentioning "`similar` for arrays and `collect` for iterable" might be simpler than introducing another `copymutable` concept.

[the remark this answers is not in the document]
Comment (LilithHafner) [reply base/sort.jl:1479]: Haha, it's piracy by OrderedCollections that allows `sort(::Dict)` to work. `OrderedCollections` just gets loaded as a transitive dependency in my startup.jl file so I didn't realize. Also the pirated `sort(::Dict)` method is deprecated.

I don't like `sort(::Dict)` returning a vector of pairs, that is rarely what the user wants and it precludes the possibility of creating a better data structure instead.

**Comment by LilithHafner** [comment]
Closing because #52098 makes the changes here obsolete.

**Pull request closed without merging**